File: dotnet/subscriptions/Customers.cs
using Pagou.Examples.Core;

namespace Pagou.Examples.Subscriptions;

// Customers are created/reused through the raw client so the subscription
// example always has a valid customer id to bill.
public static class Customers
{
    /// <summary>Reuses PAGOU_CUSTOMER_ID when set, otherwise creates a fresh customer.</summary>
    public static async Task<Customer> CreateOrReuseAsync(PagouHttpClient client)
    {
        var existing = Environment.GetEnvironmentVariable("PAGOU_CUSTOMER_ID");
        if (!string.IsNullOrEmpty(existing))
        {
            return (await client.RequestDataAsync<Customer>(new RequestParams
            {
                Method = HttpMethod.Get,
                Path = $"/v2/customers/{existing}",
            })).Data;
        }

        var input = new CreateCustomerInput
        {
            Name = "Ana Souza",
            Email = $"ana.souza+{DemoData.Now()}@example.com",
            Document = new Document { Type = "CPF", Number = "19100000000" },
            Phone = "11999990000",
            ExternalRef = $"cust_{DemoData.Now()}",
        };

        return (await client.RequestDataAsync<Customer>(new RequestParams
        {
            Method = HttpMethod.Post,
            Path = "/v2/customers",
            Body = input,
        })).Data;
    }
}

File: go/subscriptions/customers.go
package subscriptions

import (
	"fmt"
	"os"
	"time"

	"github.com/PagouAi/examples/go/internal/pagou"
)

// CreateOrReuseCustomer reuses PAGOU_CUSTOMER_ID when set, otherwise creates a
// fresh customer. Customers are demonstrated with the raw client across the
// examples.
func CreateOrReuseCustomer(client *pagou.Client) (pagou.Customer, error) {
	if existing := os.Getenv("PAGOU_CUSTOMER_ID"); existing != "" {
		res, err := pagou.RequestData[pagou.Customer](client, pagou.RequestParams{
			Method: "GET",
			Path:   fmt.Sprintf("/v2/customers/%s", existing),
		})
		return res.Data, err
	}

	now := time.Now().UnixMilli()
	input := pagou.CreateCustomerInput{
		Name:        "Ana Souza",
		Email:       fmt.Sprintf("ana.souza+%d@example.com", now),
		Document:    pagou.Document{Type: "CPF", Number: "19100000000"},
		Phone:       "11999990000",
		ExternalRef: fmt.Sprintf("cust_%d", now),
	}

	res, err := pagou.RequestData[pagou.Customer](client, pagou.RequestParams{
		Method: "POST",
		Path:   "/v2/customers",
		Body:   input,
	})
	return res.Data, err
}
